frontend/src/metabase/components/Timeline/Timeline.tsx

```diff
@@ -0,0 +1,107 @@
+import React from "react";
+import _ from "underscore";
+import { t } from "ttag";
+import { getRelativeTime } from "metabase/lib/time";
+
+import type { Revision } from "metabase-types/api";
+import Button from "metabase/core/components/Button";
+import Tooltip from "metabase/core/components/Tooltip";
+
+import {
+  TimelineContainer,
+  TimelineEvent,
+  Border,
+  StyledIcon,
+  EventBody,
+  EventHeader,
+  Timestamp,
+} from "./Timeline.styled";
+
+const ICON_SIZE = 16;
+const HALF_ICON_SIZE = ICON_SIZE / 2;
+
+type Icon = string | { name: string; color: string } | Record<string, never>;
+export type TimelineEvent = {
+  title: string;
+  timestamp: string;
+  icon: Icon;
+  description?: string;
+  revision?: Revision;
+};
+
+interface TimelineProps {
+  events: TimelineEvent[];
+  "data-testid": string;
+  canWrite: boolean;
+  revert: (revision: Revision) => void;
+  className?: string;
+}
+
+export function Timeline({
+  events,
+  "data-testid": dataTestId,
+  canWrite,
+  revert,
+  className,
+}: TimelineProps) {
+  return (
+    <TimelineContainer
+      leftShift={HALF_ICON_SIZE}
+      bottomShift={HALF_ICON_SIZE}
+      className={className}
+      data-testid={dataTestId}
+    >
+      {events.map((event, index) => {
+        const { icon, title, description, timestamp, revision } = event;
+        const isNotLastEvent = index !== events.length - 1;
+        const isNotFirstEvent = index !== 0;
+
+        return (
+          <TimelineEvent
+            key={revision?.id ?? `${title}-${timestamp}`}
+            leftShift={HALF_ICON_SIZE}
+          >
+            {isNotLastEvent && <Border borderShift={HALF_ICON_SIZE} />}
+            <EventIcon icon={icon} />
+            <EventBody>
+              <EventHeader>
+                <span>{title}</span>
+                {revision && canWrite && isNotFirstEvent && (
+                  <Tooltip tooltip={t`Revert to this version`}>
+                    <Button
+                      icon="revert"
+                      onlyIcon
+                      borderless
+                      onClick={() => revert(revision)}
+                      data-testid="question-revert-button"
+                      aria-label={t`revert to ${title}`}
+                    />
+                  </Tooltip>
+                )}
+              </EventHeader>
+              <Timestamp dateTime={timestamp}>
+                {getRelativeTime(timestamp)}
+              </Timestamp>
+              {revision?.has_multiple_changes && <div>{description}</div>}
```

Comment:
I don't think BE should return `nil` description since before this PR it's the only field, `title` is currently added.

And since this is a generic API, we might break others thing if we do that.

Reply:
Hmm. I'm trying to figure out how to avoid having a duplicate `title` and `description`, and logic distributed between the FE and BE determining what gets displayed where.

What if we kept `has_multiple_changes` but removed `title` from the API response body, and displayed `edited this.` as the title if `has_multiple_changes` is true? Otherwise the title would be the description.

Reply:
that would work! should have thought of this first, I think FE could happily handle that.

Reply:
I made the change here 5aafd02835

@EmmadUsmani I made some changes to FE to test things out, feel free to edit it if it doesn't look right for you :)

Reply:
The change looks good to me